e2e: node: rewrite cpumanager tests

ffromani opened the pull request:
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:
Rewrite the CPU Manager e2e tests, which was long in the making, because these are currently all over the place and with many known issues. The state is quite bad and a rewrite was long due.

#### Which issue(s) this PR fixes:
Partially Fixes 129884
Partially Fixes 100145

#### Special notes for your reviewer:
##### Plan and steps ahead:

1. add the new suite which overlaps with the old suite.
1. run the two suites in parallel for few weeks. Stabilize the new one, get signal
1. after 3-4 weeks the new suite should either be stable and trustworthy or proven not ready yet.
1. if not ready: revert and iterate
1. if ready: remove the old suite, push the final cleanup

##### Status report:
|content | existing tests match | done | notes |
| --- | --- | --- | --- |
|foundations, guidelines, and very basic tests | runNonGuPodTest, runGuPodTest(1), runMultipleGuNonGuPods, SMT alignment tests | :white_check_mark: | |
|strict CPU reservation tests || :white_check_mark: | |
|CFS quota tests| | :white_check_mark: | |
|option compatibility tests: full-pcpus-only and strict-cpu-reservation || :white_check_mark: | |
|option compatibility tests: full-pcpus-only and distribute-cpu-across-numa || :white_check_mark: | |
|multi-container tests |runMultipleCPUGuPod, runMultipleCPUContainersGuPod | :white_check_mark: | |
|multi-pods tests |runMultipleGuPods | :white_check_mark: | minimal coverage on par with existing tests |
|state file cleanup tests || :red_circle:  | unclear status of existing test; will re-evaluate later|
|sidecar container cpu reuse tests || :white_check_mark: ||

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

*Related* to KEP https://github.com/kubernetes/enhancements/issues/2625 - we agreed to begin and do as much as possible to the rewrite work in the context of 2625, but the KEP graduation is not blocked by the rewrite.

ffromani commented:
/sig node

ffromani commented:
/cc @swatisehgal

which was both interested and kinda waiting for this work

ffromani commented:
/test pull-kubernetes-node-kubelet-serial-containerd
/test pull-kubernetes-node-kubelet-serial-containerd-sidecar-containers
/test pull-kubernetes-node-kubelet-serial-cpu-manager
/test pull-kubernetes-node-kubelet-serial-hugepages
/test pull-kubernetes-node-kubelet-serial-memory-manager
/test pull-kubernetes-node-kubelet-serial-topology-manager

ffromani commented:
/test pull-kubernetes-node-kubelet-serial-containerd
/test pull-kubernetes-node-kubelet-serial-containerd-sidecar-containers
/test pull-kubernetes-node-kubelet-serial-cpu-manager
/test pull-kubernetes-node-kubelet-serial-hugepages
/test pull-kubernetes-node-kubelet-serial-memory-manager
/test pull-kubernetes-node-kubelet-serial-topology-manager

ffromani commented:
/test pull-kubernetes-node-kubelet-serial-containerd
/test pull-kubernetes-node-kubelet-serial-containerd-sidecar-containers
/test pull-kubernetes-node-kubelet-serial-cpu-manager
/test pull-kubernetes-node-kubelet-serial-hugepages
/test pull-kubernetes-node-kubelet-serial-memory-manager
/test pull-kubernetes-node-kubelet-serial-topology-manager

ffromani commented:
/retest

ffromani commented:
/retest-required

unrelated timeout?

ffromani commented:
/test pull-kubernetes-node-kubelet-serial-cpu-manager

ffromani commented:
/test pull-kubernetes-node-kubelet-serial-cpu-manager

ffromani commented:
/test pull-kubernetes-node-kubelet-serial-topology-manager
/test pull-kubernetes-node-kubelet-serial-topology-manager-kubetest2

ffromani committed:
e2e: node: rewrite compatibility tests

Rewrite the policy option compatibility tests.
We start with the tests which check the compatibility
between the `full-pcpus-only` and `strict-cpu-reservation`
tests, because the former is the only GA option
at time of writing.

All testcases are preserved.

Signed-off-by: Francesco Romani <fromani@redhat.com>

ffromani committed:
e2e: node: rewrite more compatibility tests

Complete the rewrite the policy option compatibility tests,
rewriting the tests which check compatibility
between the `full-pcpus-only` and `distribute-cpus-across-numa`.

All testcases are preserved.

Signed-off-by: Francesco Romani <fromani@redhat.com>

ffromani commented:
/test pull-kubernetes-node-kubelet-serial-containerd
/test pull-kubernetes-node-kubelet-serial-containerd-sidecar-containers
/test pull-kubernetes-node-kubelet-serial-cpu-manager
/test pull-kubernetes-node-kubelet-serial-hugepages
/test pull-kubernetes-node-kubelet-serial-memory-manager
/test pull-kubernetes-node-kubelet-serial-topology-manager
/test pull-kubernetes-node-kubelet-serial-topology-manager-kubetest2

ffromani commented:
/test pull-kubernetes-node-kubelet-serial-containerd

esotsal commented on test/e2e_node/cpu_manager_test.go:
```diff
@@ -923,179 +687,11 @@ func runCPUManagerTests(f *framework.Framework) {
 	f.It("should not reuse CPUs of restartable init containers", feature.SidecarContainers, func(ctx context.Context) {
 		cpuCap, cpuAlloc, _ = getLocalNodeCPUDetails(ctx, f)
 
 		// Skip rest of the tests if CPU capacity < 3.
```
```suggestion
		// Skip rest of the tests if Kubernetes CPU allocatable < 3.
```

ffromani replied:
fine point. I'm not addressing this because I'm removing all but the absolutely needed changes from `cpu_manager_test.go`, so this is not relevant in the next upload

esotsal commented on test/e2e_node/topology_manager_test.go:
```diff
@@ -680,7 +680,7 @@ func teardownSRIOVConfigOrFail(ctx context.Context, f *framework.Framework, sd *
 
 func runTMScopeResourceAlignmentTestSuite(ctx context.Context, f *framework.Framework, configMap *v1.ConfigMap, reservedSystemCPUs, policy string, numaNodes, coreCount int) {
-	threadsPerCore := getSMTLevel()
+	threadsPerCore := smtLevelFromSysFS()
```
```suggestion
	smtLevel := smtLevelFromSysFS()
```

Above will requires also update in line 717

```
numCores := smtLevel * coreCount
```

Reasoning for improved consistency, proposed to use one common variable name to return value as done in cpu_manager_test.go and cpu_manager_metrics_test.go

ffromani replied:
right call about consistency, changing as suggested

esotsal commented on test/e2e_node/topology_manager_test.go:
```diff
@@ -868,7 +868,7 @@ func runTMScopeResourceAlignmentTestSuite(ctx context.Context, f *framework.Fram
 
 func runTopologyManagerNodeAlignmentSuiteTests(ctx context.Context, f *framework.Framework, sd *sriovData, reservedSystemCPUs, policy string, numaNodes, coreCount int) {
-	threadsPerCore := getSMTLevel()
+	threadsPerCore := smtLevelFromSysFS()
```
```suggestion
	smtLevel := smtLevelFromSysFS()

```
Above will requires also update in line 914

```
numCores := smtLevel * coreCount
```
Reasoning for improved consistency, proposed to use one common variable name to return value as done in cpu_manager_test.go and cpu_manager_metrics_test.go

ffromani replied:
thanks, addressed

esotsal commented on test/e2e_node/cpu_manager_test.go:
```diff
@@ -60,6 +53,7 @@ type ctnAttribute struct {
 	cpuRequest    string
 	cpuLimit      string
 	restartPolicy *v1.ContainerRestartPolicy
+	ctnCommand    string
```
Suggestion to move this line after line 52.
Reasoning if grouped with ctnName,  helps a human mind to process faster the code while reading, avoiding context switching ctn - cpu - restartPolicy - ctn again.

ffromani replied:
fine point, I'll added at the end of the struct but doing like you suggested looks nicer indeed.

ffromani commented:
reviewers and approvers, do you prefer me to rewrite and port all the existing cpumanager in this PR or to finalize the already big present PR only addressing bugs and review comments, and finish the work with followup PRs?

esotsal commented:
I am fine with both options @ffromani, my comments are mostly are nit , besides looks good to me , your proposed refactoring is an improvement to the right direction. Thanks

ffromani commented:
I'm inclined towards rewriting all tests but keeping the old ones for a while, so we can gain trust in the rewrite before we delete it.

ffromani committed:
e2e: node: rewrite strict-cpu-reservation tests

rewrite the cpumanager e2e tests for the
`strict-cpu-reservation` policy option to fit
into the new layout.

All testcases are preserved.

Signed-off-by: Francesco Romani <fromani@redhat.com>

ffromani committed:
e2e: node: rewrite cpumanager tests

Rewrite the cpumanager tests to make use of the lessons
learned, more modern idioms, remove obsolete assumptions
and in gneeral remove all the legacy which was accumulating
over the years.

The goal is to have a simpler, flatter and more maintenable
code layout, de-entangle the net of dependency,
making the tests more robust and easier to extend.

In short, this is all about maintainability. All the testcases
will be preserved, and few other can be added along the way.

Comments in the code will explain the code layout decisions
and tradeoff, and provide a good guide to add more tests
in the future.

Special care was added in order to maximize the isolation between
tests, at cost, in selected cases of a controlled and planned
code duplication.

Signed-off-by: Francesco Romani <fromani@redhat.com>

ffromani committed:
e2e: node: rewrite CFS quota tests

Rewrite the e2e cpumanager tests about CFS quota management.
All testcases are preserved.

Signed-off-by: Francesco Romani <fromani@redhat.com>

ffromani commented:
addressed comments and only adding **intentonally duplicate** e2e tests, so we can compare the old and the new tests and gain confidence in the new ones without losing any signal. I'llproceed to  trobuleshoot the failures/flakes in the new tests.

SergeyKanzhelev commented:
/assign @swatisehgal

ffromani commented:
/test pull-kubernetes-e2e-gce
 /test pull-kubernetes-e2e-kind

ffromani commented:
/test pull-kubernetes-e2e-kind

ffromani committed:
e2e: node: rewrite: multi-cpus single-container pods

We have tests which cover the case on which a pod
with a single container require multiple CPUs;
rewrite them preserving the testcases and actually
adding coverage.

Add and use stricter checks along the way.

Signed-off-by: Francesco Romani <fromani@redhat.com>

ffromani committed:
e2e: node: initial multi-container tests

rewrite tests which exercise multiple container within the
same pod. Preserve the existing testcases, add more.

Note basic coverage for mixed pods - some containers requiring
exclusive CPUs, some not, was already added with the initial batch.

Signed-off-by: Francesco Romani <fromani@redhat.com>

ffromani committed:
e2e: node: rewrite multi-pod tests

rewrite tests which check with multiple pods.
We may extend the coverage later on.

Signed-off-by: Francesco Romani <fromani@redhat.com>

ffromani committed:
e2e: node: add comment about reserved CPU

Let's have a single general comment instead of
copypasting the same text all around the tests.

Signed-off-by: Francesco Romani <fromani@redhat.com>

ffromani committed:
e2e: node: dissolve skipIfNotEnoughAllocatableCPUs

reuse existing building blocks at the cost of
a tiny, non-nested BeforeEach (which is still OK)
and some targeted duplication.

Signed-off-by: Francesco Romani <fromani@redhat.com>

ffromani committed:
e2e: node: factor out reservedCPUs

now that we have a minimal BeforeEach (and let's keep it this way)
we can factor out the reservedCPUs setting, since it's the same
code for each testcase.

Signed-off-by: Francesco Romani <fromani@redhat.com>

ffromani committed:
e2e: node: rewrite the sidecar related tests

rewrite the tests porting to the new layout and utilities.
We may add more cases and better integration in the future.

Signed-off-by: Francesco Romani <fromani@redhat.com>

ffromani commented:
/test pull-kubernetes-node-kubelet-serial-containerd
/test pull-kubernetes-node-kubelet-serial-containerd-sidecar-containers
/test pull-kubernetes-node-kubelet-serial-cpu-manager
/test pull-kubernetes-node-kubelet-serial-hugepages
/test pull-kubernetes-node-kubelet-serial-memory-manager
/test pull-kubernetes-node-kubelet-serial-topology-manager

ffromani commented:
/test pull-kubernetes-node-kubelet-serial-cpu-manager

ffromani commented:
/test pull-kubernetes-node-kubelet-serial-containerd
/test pull-kubernetes-node-kubelet-serial-containerd-sidecar-containers
/test pull-kubernetes-node-kubelet-serial-cpu-manager
/test pull-kubernetes-node-kubelet-serial-hugepages
/test pull-kubernetes-node-kubelet-serial-memory-manager
/test pull-kubernetes-node-kubelet-serial-topology-manager

ffromani commented:
/test pull-kubernetes-e2e-gce

ffromani commented:
I can't really understand the "state file cleanup" tests: `runAutomaticallyRemoveInactivePodsFromCPUManagerStateFile`. The code doesn't match the comment, and I'm not really sure this is a good test in general. I'm inclined to leave it out from this PR and rethink it later. We will discuss in the sig-node CI meeting.

ffromani commented:
there are 2 known failures which seems to happen only in CI (can't repro locally) which I need to understand. Working on them.

esotsal commented on test/e2e_node/cpumanager_test.go:
```diff
@@ -0,0 +1,1555 @@
+					cpuLimit:   "200m",
+				},
+			})
+			pod = e2epod.NewPodClient(f).CreateSync(ctx, pod)
```
nit: To make easier troubleshooting where applicable , suggesting to add a STEP "Creating the pod" between kubelet update and checking statement

```suggestion
                        ginkgo.By("Creating the pod")
			pod = e2epod.NewPodClient(f).CreateSync(ctx, pod)
```

ffromani replied:
good suggestion, I'll add all across the board

esotsal commented on test/e2e_node/cpumanager_test.go:
```diff
@@ -0,0 +1,1555 @@
+			pod = e2epod.NewPodClient(f).CreateSync(ctx, pod)
+			podMap[string(pod.UID)] = pod
+
+			ginkgo.By("checking if the expected cpuset was assigned")
```
Perhaps adding something like this below to reduce Flaky tests? 

```suggestion
		       ginkgo.By("ensuring the testing pod is in failed state")
		       framework.ExpectNoError(e2epod.WaitTimeoutForPodReadyInNamespace(ctx, f.ClientSet, pod.Name, f.Namespace.Name, time.Minute))
			ginkgo.By("checking if the expected cpuset was assigned")
```

ffromani replied:
thing is this is baked into `CreateSync` already, so I think the cause is something else
```
// CreateSync creates a new pod according to the framework specifications, and wait for it to start and be running and ready.
func (c *PodClient) CreateSync(ctx context.Context, pod *v1.Pod) *v1.Pod {
	ginkgo.GinkgoHelper()
	p := c.Create(ctx, pod)
	framework.ExpectNoError(WaitTimeoutForPodReadyInNamespace(ctx, c.f.ClientSet, p.Name, c.namespace, framework.PodStartTimeout))
	// Get the newest pod after it becomes running and ready, some status may change after pod created, such as pod ip.
	p, err := c.Get(ctx, p.Name, metav1.GetOptions{})
	framework.ExpectNoError(err)
	return p
}
```

esotsal replied:
True

esotsal commented:
> there are 2 known failures which seems to happen only in CI (can't repro locally) which I need to understand. Working on them.

Checking the logs blob:https://prow.k8s.io/a060f0c4-3618-40aa-a63e-322eeae55f3e i suspect it is because you give very low CPUs so it take some time to start the pod? which is not been checked if it is running when checking the CPUs please check here a suggestion https://github.com/kubernetes/kubernetes/pull/130274/files#r2078001636

I notice that in other parts of the code you perform such tests so i believe this might improve the flakiness .

Good job btw :-) a joy to read this commit code, definitely improved

ffromani commented on test/e2e_node/cpumanager_test.go:
```diff
@@ -0,0 +1,1555 @@
+/*
+Copyright 2017 The Kubernetes Authors.
```
no problem, fair enough!

ffromani replied:
done

haircommander commented on test/e2e_node/topology_manager_test.go:
```diff
@@ -28,6 +27,8 @@ import (
 	"time"
 
+	"k8s.io/utils/cpuset"
+
```
nit: I Think this would be better without this space

ffromani replied:
fixed, thanks!

esotsal commented on test/e2e_node/cpumanager_test.go:
```diff
@@ -0,0 +1,1555 @@
+				{
+					ctnName:    "non-gu-container",
+					cpuRequest: "100m",
+					cpuLimit:   "200m",
```
What do you think , reasoning to test the theory that it fails in CI flake due to low cpuLimit? 

```suggestion
					cpuLimit:   "500m",
```

ffromani replied:
I'm not against this change, but the root cause turned out to be cgroup path construction done wrong. Unless the resource request is clearly(or likely) the cause for further errors/flakiness, I would prefer to defer tuning requests to followup PRs.

ffromani commented:
/test pull-kubernetes-node-kubelet-serial-containerd
/test pull-kubernetes-node-kubelet-serial-containerd-sidecar-containers
/test pull-kubernetes-node-kubelet-serial-cpu-manager
/test pull-kubernetes-node-kubelet-serial-hugepages
/test pull-kubernetes-node-kubelet-serial-memory-manager
/test pull-kubernetes-node-kubelet-serial-topology-manager

esotsal commented on test/e2e_node/cpumanager_test.go:
```diff
@@ -0,0 +1,1596 @@
+func makeCgroupPathForContainer(pod *v1.Pod, ctnName string, isInit bool) (string, error) {
+	fullCntID, ok := findContainerIDByName(pod, ctnName, isInit)
+	if !ok {
+		return "", fmt.Errorf("cannot find status for container %q", ctnName)
```
Reasoning to allow us to create targeted Kubernetes Aggregated Failures when troubleshooting flaky tests and distinguish failures between initContainers.

```suggestion
		if isInit {
		         return "", fmt.Errorf("cannot find status for init-container %q", ctnName)
		} else {
		         return "", fmt.Errorf("cannot find status for container %q", ctnName)
		}
```

ffromani replied:
fixed slightly differently than suggested

ffromani commented:
/test pull-kubernetes-node-kubelet-serial-cpu-manager

ffromani commented:
I think I'm doing the cgroup path construction wrong. I'll check how other e2e tests are doing and adapt.

ffromani commented:
/test pull-kubernetes-node-kubelet-serial-cpu-manager

ffromani commented:
yes, I'm doing it wrong: 
```
  I0508 13:59:07.858924 2617 cpumanager_test.go:133] Online CPUs: 0-31
  I0508 13:59:07.858987 2617 cpumanager_test.go:136] SMT level: 2
  I0508 13:59:07.858996 2617 cpumanager_test.go:139] cgroup V2 enabled: true
  I0508 13:59:07.859002 2617 cpumanager_test.go:140] cgroup driver: systemd
```
will change the path construction logic (ofc reusing existing code in the testsuite) and that should solve the problem

ffromani commented:
/test pull-kubernetes-node-kubelet-serial-cpu-manager

ffromani commented:
/test pull-kubernetes-node-kubelet-serial-cpu-manager

ffromani commented:
/test pull-kubernetes-node-kubelet-serial-containerd
/test pull-kubernetes-node-kubelet-serial-containerd-sidecar-containers
/test pull-kubernetes-node-kubelet-serial-cpu-manager
/test pull-kubernetes-node-kubelet-serial-hugepages
/test pull-kubernetes-node-kubelet-serial-memory-manager
/test pull-kubernetes-node-kubelet-serial-topology-manager

ffromani commented:
/test pull-kubernetes-node-kubelet-serial-containerd
/test pull-kubernetes-node-kubelet-serial-containerd-sidecar-containers
/test pull-kubernetes-node-kubelet-serial-cpu-manager
/test pull-kubernetes-node-kubelet-serial-hugepages
/test pull-kubernetes-node-kubelet-serial-memory-manager
/test pull-kubernetes-node-kubelet-serial-topology-manager

ffromani commented:
I cleaned up the fixing commits. There are no known issues anymore (last famous words?). The PR is reviewable again

more testing using pre-submit jobs is welcome. Which other presumbit job may be relevant?

ffromani commented:
/test pull-kubernetes-node-kubelet-serial-containerd

ffromani commented:
/test pull-kubernetes-node-kubelet-serial-containerd

```
E2eNode Suite: [It] [sig-node] Device Plugin [Serial] [Feature:DevicePlugin] DevicePlugin [Serial] [Disruptive] Does not keep device plugin assignments across node reboots if fails admission (no pod restart, no device plugin re-registration)
```
device plugin flakiness, will address in another PR

ffromani commented:
/triage accepted
/priority important-longterm

esotsal commented:
> /test pull-kubernetes-node-kubelet-serial-containerd
> 
> ```
> E2eNode Suite: [It] [sig-node] Device Plugin [Serial] [Feature:DevicePlugin] DevicePlugin [Serial] [Disruptive] Does not keep device plugin assignments across node reboots if fails admission (no pod restart, no device plugin re-registration)
> ```
> 
> device plugin flakiness, will address in another PR

Yes, flaky test is not related with your commit https://github.com/kubernetes/kubernetes/issues/131703 

/lgtm

ffromani commented:
> > device plugin flakiness, will address in another PR
> 
> Yes, flaky test is not related with your commit , opened #131703
> 
> /lgtm

thanks, will look at this separately.
